### ISSUE TITLE
Weird hash comparision on backblaze b2

### ISSUE THREAD

**Issue by NoSyu** [ncw/rclone#578]
When filing an issue, please include the following information if
possible as well as a description of the problem.

> What is your rclone version (eg output from `rclone -V`)
> rclone v1.30
> 
> Which OS you are using and how many bits (eg Windows 7, 64 bit)
> Windows 7, 64 bit
> 
> Which cloud storage system are you using? (eg Google Drive)
> Backblaze b2
> 
> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
> rclone sync D:\backup\Personal\ remote:NoSyuPersonalBackUP --transfers 100 --checksum -n --retries 50 -v --log-file NoSyuPersonalBackUP_%ldt%.log
> 
> A log from the command with the `-v` flag (eg output from `rclone -v copy /tmp remote:tmp`)
> 2016/07/10 21:56:38 video/20150830 MLSS 2015 Video.zip: Hash differ
> 2016/07/10 21:56:38 video/20150830 MLSS 2015 Video.zip: Not copying as --dry-run

Hello
I try to sync the local files to backblaze b2 with rclone. Most of the files are copied correctly, but some files occur weird errors.

The error messages of the files are 'Hash differ'.
So, I checked the hash value of the files manually by 'rclone sha1sum' operation.
The results are as below:

> E:\rclone-v1.30-windows-amd64>rclone.exe sha1sum "D:\backup\Personal\video\20150830 MLSS 2015 Video.zip"
> dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f  20150830 MLSS 2015 Video.zip
> 2016/07/10 21:40:15
> Transferred:            0Bytes (0Byte/s)
> Errors:                 0
> Checks:                 1
> Transferred:            0
> Elapsed time:       50.9s
> 
> E:\rclone-v1.30-windows-amd64>rclone.exe sha1sum "remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip"
> dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f  20150830 MLSS 2015 Video.zip
> 2016/07/10 21:41:06
> Transferred:            0Bytes (0Byte/s)
> Errors:                 0
> Checks:                 1
> Transferred:            0
> Elapsed time:        2.2s
> 
> E:\rclone-v1.30-windows-amd64>rclone.exe size "D:\backup\Personal\video\20150830 MLSS 2015 Video.zip"
> Total objects: 1
> Total size: 5.271G (5659957144 bytes)
> 2016/07/10 21:43:00
> Transferred:            0Bytes (0Byte/s)
> Errors:                 0
> Checks:                 0
> Transferred:            0
> Elapsed time:           0
> 
> E:\rclone-v1.30-windows-amd64>rclone.exe size "remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip"
> Total objects: 1
> Total size: 5.271G (5659957144 bytes)
> 2016/07/10 21:42:45
> Transferred:            0Bytes (0Byte/s)
> Errors:                 0
> Checks:                 0
> Transferred:            0
> Elapsed time:        2.8s

The local and remote files have same sha1 value, and the size of the files are also same. But, when I run the sync, it shows 'hash differ'.

I am checking the source code, and test it. If I found the bug or any enhancement, I will commit the code.

Thanks
Yours sincerely

**Comment by ncw** [comment]
I suspect this was fixed in e0aa4bb4923237585dc301f41b1f799c1b7cd564 as part of #533 

Have a go with the latest beta: http://pub.rclone.org/v1.30-28-g2a1d4b7%CE%B2/

Thanks

Nick

**Comment by NoSyu** [comment]
Thank you!
I check the latest beta program, and it works fine.

In fact, to avoid the issue #533, I also made own rclone following the instruction (http://rclone.org/install/) at 2016/07/09, and tested. But the program produces same error. So I wrote this article.
I think my installation have a problem. I re-check the process.

Thanks
Yours sincerly

**Comment by ncw** [comment]
@NoSyu thanks for testing.  Not sure what went wrong with your own compile - the code should have been up to date on github.  Did you do `go get -u -v github.com/ncw/rclone/...`?

**Comment by NoSyu** [comment]
@ncw I recompile the code, and test it. The result is OK, no problem.
Sorry for my mistake.

**Comment by ncw** [comment]
@NoSyu Thanks for checking!

**Comment by NoSyu** [comment]
Sorry for re-opening this issue, because this problem appears at v1.32 in Windows amd64 version, with same file - 20150830 MLSS 2015 Video.zip.
To check the problem in more detail, I add a code to print the hash of source file and destination file when the 'Hash differ' message is printed based on c2f6decb9cbd8b8bf504fdac76a9f51d288832fd commit in the function CheckHashes in fs/operations.go file.

The result is as below (First is the source file in local disk and second is the destination file in B2):

> 2016/07/24 12:00:49 video/20150830 MLSS 2015 Video.zip: dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f
2016/07/24 12:00:49 video/20150830 MLSS 2015 Video.zip: 9562fd50e2d953c889c95423c8357970fba44f8c
2016/07/24 12:00:49 video/20150830 MLSS 2015 Video.zip: Hash differ
2016/07/24 12:00:49 video/20150830 MLSS 2015 Video.zip: Not copying as --dry-run

After that, I run the 'sha1sum' operation, and the result is not the same above.

> E:\rclone>rclone.exe sha1sum "remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip"
dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f  20150830 MLSS 2015 Video.zip
2016/07/24 12:32:26
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 1
Transferred:            0
Elapsed time:        2.6s

I also check the hash file in B2 webpage.

>large_file_sha1: dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f   

I think this is a bug. Please check the issue one more time.

Thank you

**Comment by ncw** [comment]
I think this is a different issue you are right.

So what you seem to be saying is when you do `rclone --checksum copy src b2:` the checksums don't match with the one being read from b2: being the incorrect one.

Can you try this for me which should attempt to copy just the one file (with `--dry-run`) that is causing a problem.  It will make a lot of log messages into `log1.txt` - can you remove the `Authorization:` lines and attach them to a post (or email them to me if you prefer).

    rclone.exe --dry-run -v --dump-headers copy  --include "/20150830 MLSS 2015 Video.zip" "D:\backup\Personal\video" "remote:NoSyuPersonalBackUP/video/" --log-file log1.txt

Can you also do the same with this command and `log2.txt`

    rclone.exe -v --dump-headers sha1sum "remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip" --log-file log2.txt

That should tell me exactly what is going on.

Thanks

Nick

**Comment by ncw** [comment]
Correction - make those `--dump-headers` into `--dump-bodies` (have edited the post)

**Comment by NoSyu** [comment]
I re-re-re test the this issue, and the problem appears only when I set the path to root as below:

> rclone sync D:\backup\Personal\ remote:NoSyuPersonalBackUP -n --transfers 32 --checksum --retries 1000 -v --log-file NoSyuPersonalBackUP_%ldt%.log

When I set the path to the specific directory, then it works fine.

> rclone sync D:\backup\Personal\video remote:NoSyuPersonalBackUP/video -n --transfers 32 --checksum --retries 1000 -v --log-file NoSyuPersonalBackUP_%ldt%.log

The result of the last operation is as below:
> 2016/07/24 18:07:38 20150830 MLSS 2015 Video.zip: dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f
2016/07/24 18:07:38 20150830 MLSS 2015 Video.zip: dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f
2016/07/24 18:07:38 20150830 MLSS 2015 Video.zip: Size and SHA-1 of src and dst objects identical
2016/07/24 18:07:38 20150830 MLSS 2015 Video.zip: Unchanged skipping

I also check it with v1.32, and there is no hash differ with the last execution.

I will try to the operations as you mentioned, and check the results.

Thank you

**Comment by ncw** [comment]
I wonder if on b2 you have two objects with differing case, eg

    remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip
    remote:NoSyuPersonalBackUP/Video/20150830 MLSS 2015 Video.zip

That might explain what is going on.

**Comment by NoSyu** [comment]
I run the code which has `-dump-bodies` command. However, there is no error in both cases.
As I mentioned https://github.com/ncw/rclone/issues/578#issuecomment-234771753, the error only appears when I set the root path. 
So I run the `-dump-bodies` command to root path. It takes long time. I will update the information ASAP.

**Comment by ncw** [comment]
OK thanks.

**Comment by NoSyu** [comment]
Thank you for suggsting the reason of the issue. However it's not.

I check the filename as below:

>E:\rclone>rclone ls "remote:NoSyuPersonalBackUP/video/20150830 MLSS 2015 Video.zip"
5659957144 20150830 MLSS 2015 Video.zip
2016/07/24 20:56:48
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:        3.6s

>E:\rclone>rclone ls "remote:NoSyuPersonalBackUP/Video/20150830 MLSS 2015 Video.zip"
2016/07/24 20:56:58
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:        3.1s

>E:\rclone>rclone ls "D:\backup\Personal\video\20150830 MLSS 2015 Video.zip"
5659957144 20150830 MLSS 2015 Video.zip
2016/07/24 20:57:53
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:           0

>E:\rclone>rclone ls "D:\backup\Personal\Video\20150830 MLSS 2015 Video.zip"
5659957144 20150830 MLSS 2015 Video.zip
2016/07/24 20:58:00
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:           0

![qqqq](https://cloud.githubusercontent.com/assets/2724705/17083699/6871f38a-51d2-11e6-9286-8a967c32598c.png)

The location of the file are in `video` directory in both local and remote.

I also re-check it with ls operator to B2 remote root path, and there is only one `20150830 MLSS 2015 Video.zip` file in the bucket.

**Comment by ncw** [comment]
OK thanks for checking

**Comment by NoSyu** [comment]
Here I attached the snipped logs which contain the `20150830 MLSS 2015 Video.zip`. 

>{
      "action": "upload",
      "contentLength": 5659957144,
      "contentSha1": null,
      "contentType": "application/octet-stream",
      "fileId": "4_zcc94124fb690a616565d071f_f2004db74d996abe3_d20160723_m171923_c001_v0001004_t0013",
      "fileInfo": {
        "large_file_sha1": "dd494d689cceac5e3e62e17f8a1ae2ab67eeea4f",
        "src_last_modified_millis": "1467909807201"
      },
      "fileName": "video/20150830 MLSS 2015 Video.zip",
      "size": 5659957144,
      "uploadTimestamp": 1469294363000
    },

>2016/07/24 21:17:57 video/20150830 MLSS 2015 Video.zip: Hash differ
2016/07/24 21:17:57 video/20150830 MLSS 2015 Video.zip: Not copying as --dry-run

I run the v1.32 windows amd64 program.

I will send the whole file via email, because it has all filenames in my personal back directory.

Thank you very much

**Comment by ncw** [comment]
Thanks for that.  The return from B2 looks disappointingly normal.

In your debugging are you sure it was the B2 shasum that was wrong - could it have been the local one?

I will study your log some more.

**Comment by NoSyu** [comment]
```
func CheckHashes(src, dst Object) (equal bool, hash HashType, err error) {
	common := src.Fs().Hashes().Overlap(dst.Fs().Hashes())
	// Debug(nil, "Shared hashes: %v", common)
	if common.Count() == 0 {
		return true, HashNone, nil
	}
	hash = common.GetOne()
	srcHash, err := src.Hash(hash)
	if err != nil {
		Stats.Error()
		ErrorLog(src, "Failed to calculate src hash: %v", err)
		return false, hash, err
	}
	if srcHash == "" {
		return true, HashNone, nil
	}
	dstHash, err := dst.Hash(hash)
	if err != nil {
		Stats.Error()
		ErrorLog(dst, "Failed to calculate dst hash: %v", err)
		return false, hash, err
	}
	if dstHash == "" {
		return true, HashNone, nil
	}

	Debug(src, srcHash)
	Debug(dst, dstHash)

	return srcHash == dstHash, hash, nil
}
```

This is a code which prints the hash values at https://github.com/ncw/rclone/issues/578#issuecomment-234755514
Only the changes is calling two Debug function with srcHash and dstHash.
I wll also investigate this problem in more detail.

**Comment by ncw** [comment]
Are you still having a problem with this?

Have you tried the latest v1.33?

**Comment by NoSyu** [comment]
Yes, the problem is not solved with v.1.33 windows x64. So I solved this problem by divide and conquer - means run rclone sync with sub directories. :)
Now, I am out of the office and home, and come back to at 5th Sep. After that, I re-check this problem in more detail since the structure of the rclone is changed a lot.
Thank you

**Comment by ncw** [comment]
Great thanks. The more help pinning it down the better.

**Comment by pnoc** [comment]
Hi

I have the same issue with rclone v1.34 (Ubunutu 16.10 64bits) and backblaze...
Rclone always uploads the same sets of file over and over on  `rclone copy --checksum` commands although the sha1 is identical (i checked locally with sha1sum, it matches b2's large_file_sha1 field).
It only seems to occur on large files (>200 Mb), could it be an issue with the file chunking? 
Actually, I have a few folder where every large files have the issue (but not the small files)...
It doesn't happen in every folder containing large files though, just a few ones (that's quite weird).

I hope this can help!
In the meantime, I'll remove the checksum option during my backups and I will run the "check" command later (I don't know if it would work better)...

(PS : thank you for this amazing tool! keep up the good work! :-) )

**Comment by pnoc** [comment]
Same issue with rclone check --checksum

I did some additional tests with --dump-bodies flag and I added a debug message in CheckHashes function with the 2 compared hashes.
The "srcHash" is the right one, but the "dstHash" is sometime completly wrong!
I searched for the value of the dstHash in the dump-bodies part of the file, it matches another file's SHA1 in a different folder.
Could it be a concurrency issue, which could explain this file/hash mix-up ? or something wrong during the lookup of the remote sha1? all remote requests seems to be executed first, how are the results stored/indexed/accessed (i'm not familiar with the go language)  ?

I tried with --checkers=1, but the issue is still there (I may have a lower error rate, I didn't make a full run, since it's quite slow)...

I did try to run the command in one of the offending sub-directory: the checksum verification is correct... The issue only appears on larger sync / checks, so it may be hard to reproduce.
My backup consists of 60000+ files and a total of 500+ Gb of data (no very large files, file size are mostly lower than 1Gb)

A few logs that demonstrates my findings : 

2017/01/08 00:24:19 d/TY/BQQX2EUOLXQPB6YGWU4WCBRIEIEO5I/NU2VKDWICTF4CWGHRE6WXMFUJARHA2GKIDK3HP3RYNNSGV2NL4JTVW6P4GO64===: ##### DEBUG#### srcHash= 5b76d5aa3bf065e123e0c1b46c9245de1c2b39e3 ---- destHash= 680294244c3bc820160fd8b065e37e5a453e03e6
2017/01/08 00:24:19 d/TY/BQQX2EUOLXQPB6YGWU4WCBRIEIEO5I/NU2VKDWICTF4CWGHRE6WXMFUJARHA2GKIDK3HP3RYNNSGV2NL4JTVW6P4GO64===: SHA-1 differ

--dump-bodies match

    {
      "action": "upload",
      "contentLength": 200200264,
      "contentSha1": null,
      "contentType": "application/octet-stream",
      "fileId": "XXXXXXXXXXX",
      "fileInfo": {
        "large_file_sha1": "5b76d5aa3bf065e123e0c1b46c9245de1c2b39e3",
        "src_last_modified_millis": "1482950502875"
      },
      "fileName": "d/TY/BQQX2EUOLXQPB6YGWU4WCBRIEIEO5I/NU2VKDWICTF4CWGHRE6WXMFUJARHA2GKIDK3HP3RYNNSGV2NL4JTVW6P4GO64===",
      "size": 200200264,
      "uploadTimestamp": 1482961087000
    },
    {
      "action": "upload",
      "contentLength": 2515656,
      "contentSha1": "680294244c3bc820160fd8b065e37e5a453e03e6",
      "contentType": "application/octet-stream",
      "fileId": "XXXXXXXXXXXXX",
      "fileInfo": {
        "src_last_modified_millis": "1482178222189"
      },
      "fileName": "d/TH/T4FL3DASV6EF34GD2W4HBLIXDMNFH6/JNMWI5UEMCMRPGRE5E6MKADCOPL7RJ5DYTEUUSDYSEWCG===",
      "size": 2515656,
      "uploadTimestamp": 1482268518000
    },

(Note : my data is locally encrypted by cryptomator, hence the weird obfuscated filenames)

The confusion always seems to be between a "large_file_sha1" field of one file and a "contentSha1" of a small file...

**Comment by ncw** [comment]
> The confusion always seems to be between a "large_file_sha1" field of one file and a "contentSha1" of a small file...

Strange!

Did you find the corrupted hash anywhere else - maybe in the file headers?

Can you try again with the latest beta: http://beta.rclone.org/v1.35-19-g499766f/ - I did change the B2 directory iteration in 1.35 to simplify it a lot.

The rclone check is a good test to do.

If that doesn't help, then what we need to do is build rclone with the race detector to see if there is some concurrency problem in rclone.  Unfortunately I don't have a 64 bit windows machine to build it on (normally I cross compile, but you can't easily cross compile if you are using the race detector).

If you wanted to try this, then you would download go from golang.org, Set your PATH as described [here](https://golang.org/doc/install#testing), then issue the command `go get -u github.com/ncw/rclone`.  Then `go build -race github.com/ncw/rclone` which should build the rclone binary in the current directory with the race detector.  You use it as normal (run the check) and it will complain loudly to stderr if there are any races detected.

**Comment by mlbarrow** [comment]
I just tried the beta and it still has the problem that I have been experiencing. I am trying to push up a bunch of media files to B2 and strangely, it is failing on the exact same files each time. Is there anything I can do to gather more details to diagnose the issue?

**Comment by ncw** [comment]
I'm confident that this was fixed by 1d42a343d25b4ae6263fe2046c681f54ebee53dc

Please try a latest beta and re-open if you still see the problem!

https://beta.rclone.org/v1.36-41-gb651784/

Thanks

Nick
